### PR TITLE
Feature #35 Implementation - Tokens

### PR DESCRIPTION
**Pull request by Azkel**
Implementation of feature #35 , based on some decisions made within that feature discussion.

- Possibility to add token with customized label number/size/color from menu
- Right-click on the token when Token tool is selected is removing the top token where the mouse pointer is currently
- Added react-color library to provide color picker functionality for tokens

### CODE REVIEW

**Comment by apclary** [comment]
Is this ready to be merged?

**Comment by Azkel** [comment]
Yes, all is done.

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "junk": "3.1.0",
     "morgan": "~1.9.1",
     "react": "16.9.0",
+    "react-color": "^2.17.3",
```
Comment (n1ru4l) [inline package.json]: I would recommend pinning the version

File: src/dm-area/dm-map.js [hunk 2]
```diff
@@ -540,6 +555,20 @@ export const DmMap = ({
     const socket = io();
     socketRef.current = socket;
 
+    socket.on("add token", async data => {
+      setTokens(tokens => tokens.filter(area => area.id !== data.id));
```
Comment (n1ru4l) [inline src/dm-area/dm-map.js]: The two setToken calls can be combined into one

File: src/dm-area/dm-map.js [hunk 3]
```diff
@@ -673,14 +706,56 @@ export const DmMap = ({
...
-          if (tool !== "mark") {
-            return;
+        onContextMenu={ev => {
+          if (tool === "tokens") {
+            ev.preventDefault();
+            const ref = new Referentiel(
+              panZoomRef.current.dragContainer.current
+            );
+            const [x, y] = ref.global_to_local([ev.pageX, ev.pageY]);
+            const token = [...document.querySelectorAll(".tokenCircle")].filter(
```
Comment (n1ru4l) [inline src/dm-area/dm-map.js]: I don‘t like this bail out of react solution. Can we add the event listener to the token, so we do not have to manually read from the DOM?

File: src/player-area.js [hunk 4]
```diff
@@ -84,6 +85,24 @@ export const PlayerArea = () => {
       ]);
     });
 
+    socket.on("add token", async data => {
+      setTokens(tokens => tokens.filter(area => area.id !== data.id));
```
Comment (n1ru4l) [inline src/player-area.js]: These setTokens calls can be combined into one

File: src/dm-area/dm-map.js (hunk 3 above)
Comment (Azkel) [reply src/dm-area/dm-map.js]: I had an issue with this one for a while - basically as all the event listeners are on the PanZoom, I was not able to properly bind onClick/onContextMenu to the ObjectLayer.  Implementing #148 may make this workaround obsolete, but for the moment it was the best I could think of that still used event listeners on the PanZoom.

File: package.json (hunk 1 above)
Comment (Azkel) [reply package.json]: Changed in 87b4393.

File: src/player-area.js (hunk 4 above)
Comment (Azkel) [reply src/player-area.js]: Changed in 2ab34f9.

File: src/dm-area/dm-map.js (hunk 2 above)
Comment (Azkel) [reply src/dm-area/dm-map.js]: Changed in 2ab34f9.

File: src/dm-area/dm-map.js (hunk 3 above)
Comment (n1ru4l) [reply src/dm-area/dm-map.js]: Okay, I can understand this! I fixed #148 as part of #146. My current plan is to merge #146 first and then rebase your work on top of it (while adjusting the missing bits). Hopefully, that will be done until the end of the week.

Comment (Azkel) [reply src/dm-area/dm-map.js]: So, seeing as #146 has been merged down, shall I leave merging my changes to you @n1ru4l?

**Comment by n1ru4l** [comment]
Okay, I merged  #146 and will release v1.3.0 shortly. Unfortunately (for this pull request), I will be on vacation for the next two weeks. I might be able to rebase tomorrow but I cannot give you a guarantee. Will definitely prioritize this before merging other new features!

**Comment by Azkel** [comment]
I can try to rebase my changes with things from #146 myself this evening (CEST), but I'll also be on vacation starting from 11th to the last day of September - so in case of any review feedback, I'll implement it after my return.

**Comment by Azkel** [comment]
Synced with master branch, but still not done anything in terms of request change above. I'm not sure if I'll have time to modify this before my holidays unfortunatelly.

**Comment by n1ru4l** [comment]
Gonna check/test this on the weekend!

**Comment by n1ru4l** [comment]
I am doing some work on this in https://github.com/Azkel/dungeon-revealer/pull/1

File: src/dm-area/dm-map.js (hunk 3 above)
Comment (n1ru4l) [reply src/dm-area/dm-map.js]: Yes I am currently working on this!

**Comment by n1ru4l** [comment]
Let's continue this in #180!

**Pull request closed without merging**